### PR TITLE
discord: add ability to use custom applications for Rich Presence

### PR DESCRIPTION
**Pull request by vecchiotom**
this is useful for people who want to change fivem's rich presence image, it can be set from scripts using natives and also from the CitizenFX.ini file (DiscordId and DiscordImage)

P.S. this is my real first time doing anything with cpp, i have probably overcomplicated some things, if you'd like for me to change anything, just review the commit and i will try my best to fix it :)

### CODE REVIEW

File: code/components/discord/src/DiscordPresence.cpp [hunk 1]
```diff
@@ -1,4 +1,7 @@
 #include <StdInc.h>
+#include <windows.h>
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: ... why? These header files are included by StdInc anyway.

File: code/components/discord/src/DiscordPresence.cpp [hunk 2]
```diff
@@ -16,8 +19,14 @@ static std::string g_richPresenceValues[8];
 
 static std::string g_richPresenceOverride;
 
+static std::string g_richPresenceOverrideAsset;
+
 static time_t g_startTime = time(nullptr);
 
+static std::string DiscordId;
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Naming conventions, perhaps `g_discordAppId`?

File: code/components/discord/src/DiscordPresence.cpp [hunk 3]
```diff
@@ -40,13 +49,20 @@ static void UpdatePresence()
...
+		
 		DiscordRichPresence discordPresence;
 		memset(&discordPresence, 0, sizeof(discordPresence));
 		discordPresence.state = line1.c_str();
 		discordPresence.details = line2.c_str();
 		discordPresence.startTimestamp = g_startTime;
-		discordPresence.largeImageKey = "fivem_large";
+		if (!g_richPresenceOverrideAsset.empty())
+		{
+		discordPresence.largeImageKey = g_richPresenceOverrideAsset.c_str();
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: where's the indentation?

File: code/components/discord/src/DiscordPresence.cpp [hunk 4]
```diff
@@ -57,7 +73,41 @@ static InitFunction initFunction([]()
 {
 	DiscordEventHandlers handlers;
 	memset(&handlers, 0, sizeof(handlers));
-	Discord_Initialize("382624125287399424", &handlers, 1, nullptr);
+	std::wstring fpath = MakeRelativeCitPath(L"CitizenFX.ini");
+	
+	if (GetFileAttributes(fpath.c_str()) == INVALID_FILE_ATTRIBUTES)
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: INI overrides are likely undesirable. This code is also extremely ugly. 😕

File: code/components/discord/src/DiscordPresence.cpp [hunk 5]
```diff
@@ -57,7 +73,41 @@ static InitFunction initFunction([]()
...
+		wchar_t path[512];
+		const wchar_t* pathKey = L"DiscordId";
+		GetPrivateProfileString(L"Game", pathKey, NULL, path, _countof(path), fpath.c_str());
+		if (wcscmp(path, L"") == 0)
+		{
+			DiscordId = "382624125287399424";
+		}
+		else {
+			char DiscordIdS[512];
+			wcstombs(DiscordIdS, path, sizeof(DiscordIdS));
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: ... no, no, no, no, use ToWide/ToNarrow.

File: code/components/discord/src/DiscordPresence.cpp [hunk 6]
```diff
@@ -57,7 +73,41 @@ static InitFunction initFunction([]()
...
+			DiscordImage = "fivem_large";
+		}
+		else {
+			char DiscordImageS[512];
+			wcstombs(DiscordImageS, path, sizeof(DiscordImageS));
+			DiscordImage = DiscordImageS;
+		}
+		
+	}
+	Discord_Initialize(DiscordId.c_str(), &handlers, 1, nullptr);
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Variable naming, and a padding newline?

File: code/components/discord/src/DiscordPresence.cpp [hunk 7]
```diff
@@ -108,4 +158,40 @@ static InitFunction initFunction([]()
 
 		g_richPresenceChanged = true;
 	});
+
+	fx::ScriptEngine::RegisterNativeHandler("SET_RICH_PRESENCE_ASSET", [](fx::ScriptContext& context)
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Service-specific, so make the native like `SET_DISCORD_RICH_PRESENCE_ASSET`. Also, you forgot to update `codegen_cfx_natives.lua`?

File: code/components/discord/src/DiscordPresence.cpp [hunk 8]
```diff
@@ -108,4 +158,40 @@ static InitFunction initFunction([]()
...
+
+		if (str)
+		{
+			DiscordId = str;
+		}
+		else
+		{
+			DiscordId = "382624125287399424";
+		}
+		Discord_Shutdown();
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Is this call actually safe? Also, probably should be called `SET_DISCORD_APPLICATION_ID` or so.

File: code/components/discord/src/DiscordPresence.cpp [hunk 9]
```diff
@@ -57,7 +73,41 @@ static InitFunction initFunction([]()
 {
 	DiscordEventHandlers handlers;
 	memset(&handlers, 0, sizeof(handlers));
-	Discord_Initialize("382624125287399424", &handlers, 1, nullptr);
+	std::wstring fpath = MakeRelativeCitPath(L"CitizenFX.ini");
+	
+	if (GetFileAttributes(fpath.c_str()) == INVALID_FILE_ATTRIBUTES)
+	{
+		DiscordId = "382624125287399424";
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Default ID is mentioned so often it probably should be a `#define`.

File: code/components/discord/src/DiscordPresence.cpp (hunk 8 above)
Comment (vecchiotom) [reply code/components/discord/src/DiscordPresence.cpp]: according to discord's docs, it is, it terminates the thread allowing the application to shut down gracefully or discord rich presence to be reinitialized, without this, i am pretty sure the threads would just stack one onto another every time you changed the app id.

**Comment by bladecoding** [comment]
@vecchiotom You should use convars instead of reading from an INI file.

**Comment by vecchiotom** [comment]
we are talking about the client tho, not the server, and the INI file is CitizenFX.ini, which to my knowledge is what is used for client side configuration 😄

**Comment by bladecoding** [comment]
Is this meant to be user editable though? I was assuming this is so that servers can show off that you are playing on their server with a custom icon and name. If that's the case then a serverinfo convar would work best since that's sent when you connect(See: NetLibrary::ProcessOOB).

**Comment by vecchiotom** [comment]
let's see what i can do 😄

**Comment by vecchiotom** [comment]
@bladecoding still trying to find out how to get a convar value on the client, mind giving me a little clue? please 🙏 😄

**Comment by vecchiotom** [comment]
ok, pretty confident everything you asked for has now been reviewed and fixed 😄 @blattersturm

File: ext/natives/codegen_cfx_natives.lua [hunk 10]
```diff
@@ -1938,3 +1938,27 @@ native "GET_HASH_KEY"
...
 	]]
+
+native "SET_DISCORD_APP_ID"
+	arguments {
+		charPtr "appId",
+	}
+	apiset 'client'
+	returns	"void"
+	doc [[!
+<summary>
```
Comment (blattersturm) [inline ext/natives/codegen_cfx_natives.lua]: No definition of `<param />`?

File: ext/natives/codegen_cfx_natives.lua [hunk 11]
```diff
@@ -1938,3 +1938,27 @@ native "GET_HASH_KEY"
...
+	]]
+
+native "SET_DISCORD_RICH_PRESENCE_ASSET"
+	arguments {
+		charPtr "assetName",
+	}
+	apiset 'client'
+	returns	"void"
+	doc [[!
+<summary>
```
Comment (blattersturm) [inline ext/natives/codegen_cfx_natives.lua]: Again no param definitions. Also, the descriptions are a little vague and non-specific.

File: code/components/discord/src/DiscordPresence.cpp [hunk 12]
```diff
@@ -41,12 +51,21 @@ static void UpdatePresence()
...
 		memset(&discordPresence, 0, sizeof(discordPresence));
 		discordPresence.state = line1.c_str();
 		discordPresence.details = line2.c_str();
 		discordPresence.startTimestamp = g_startTime;
-		discordPresence.largeImageKey = "fivem_large";
+		if (!g_richPresenceOverrideAsset.empty())
+		{
+			discordPresence.largeImageKey = g_richPresenceOverrideAsset.c_str();
+		}
+		else {
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: { not on a new line? :(

File: code/components/discord/src/DiscordPresence.cpp [hunk 13]
```diff
@@ -41,12 +51,21 @@ static void UpdatePresence()
 			line1 = g_richPresenceOverride;
 		}
 
+
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Random added whitespace!

File: code/components/discord/src/DiscordPresence.cpp [hunk 14]
```diff
@@ -89,6 +111,8 @@ static InitFunction initFunction([]()
 	OnKillNetworkDone.Connect([]()
 	{
 		g_richPresenceOverride = "";
+		NativeInvoke::Invoke<HashString("SET_DISCORD_APP_ID"),char*,char*>(DEFAULT_APP_ID);
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Unsure whether it's a good idea to call natives inline like that from here.

File: code/components/discord/src/DiscordPresence.cpp [hunk 15]
```diff
@@ -108,4 +132,41 @@ static InitFunction initFunction([]()
...
+		if (str)
+		{
+			g_discordAppId = str;
+		}
+		else
+		{
+			g_discordAppId = DEFAULT_APP_ID;
+		}
+		
+		Discord_Shutdown();
```
Comment (blattersturm) [inline code/components/discord/src/DiscordPresence.cpp]: Can this not be done in the tick routine using a appid != lastappid check (or similar)?

File: code/components/discord/src/DiscordPresence.cpp (hunk 14 above)
Comment (vecchiotom) [reply code/components/discord/src/DiscordPresence.cpp]: yeah the idea was to avoid repetitions in the code, but sure, i can fix this :)

**Comment by vecchiotom** [comment]
pushed some changes to address the issues as requested :)

**Pull request merged**